libs/json/encoder.go

```diff
@@ -0,0 +1,246 @@
+package json
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"strconv"
+	"time"
+)
+
+var (
+	timeType            = reflect.TypeOf(time.Time{})
+	jsonMarshalerType   = reflect.TypeOf(new(json.Marshaler)).Elem()
+	jsonUnmarshalerType = reflect.TypeOf(new(json.Unmarshaler)).Elem()
+)
+
+// Marshal marshals the value as JSON, using Amino-compatible JSON encoding (strings for
+// 64-bit numbers, and type wrappers for registered types).
+func Marshal(v interface{}) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	err := encode(buf, v)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// MarshalIndent marshals the value as JSON, using the given prefix and indentation.
+func MarshalIndent(v interface{}, prefix, indent string) ([]byte, error) {
+	bz, err := Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+	buf := new(bytes.Buffer)
+	err = json.Indent(buf, bz, prefix, indent)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func encode(w io.Writer, v interface{}) error {
+	// Bare nil values can't be reflected, so we must handle them here.
+	if v == nil {
+		return writeStr(w, "null")
```

Comment:
I think it might be important to document either somewhere prominently when `null` is written vs. for instance fields being omitted instead. 

I think this is particularly relevant as some languages (e.g. rust) do not really have a `nil` equivalent (you could argue that `[None](https://doc.rust-lang.org/std/option/enum.Option.html#variant.None)` is similar but it is semantically quite different).